### ISSUE TITLE
strip out auto generated links by the browser

### ISSUE THREAD

**Issue by MrMontesa** [zammad/zammad#2019]
As per discussion on [community board](https://community.zammad.org/t/malformed-url-links-in-triggered-mails/666), I file this ticket.

### Infos:

* Used Zammad version: 2.4.0
* Installation method (source, package, ..): CentOS7 package
* Operating system: CentOS 7.4
* Database + version: postgresql 9.2.23
* Elasticsearch version: 5.6.8
* Browser + version: Chrome 66 / Outlook 2016

### Expected behavior:

* HTML Links should point to the correct destination

### Actual behavior:

* As an example the Link: _https://support.domain.com/#ticket/zoom/123_ points to **https://support.domain.com/https://support.domain.com/#ticket/zoom/123**

**Comment by martini** [comment]
For some reasons the browser is quoting ticket.id als url (I guess the browser means .id as an url). The trigger content looks like:

```
> <a href="#{config.http_type}://#{config.fqdn}/#ticket/zoom/#{<a href="#{config.http_type}://#{config.fqdn}/#ticket/zoom/#{ticket.id}">ticket.id</a>}">#{config.http_type}://#{config.fqdn}/#ticket/zoom/#{<a href="#{config.http_type}://#{config.fqdn}/#ticket/zoom/#{ticket.id}">ticket.id</a>}</a>
```

Remark: 

Wrong is:
`#{<a href="#{config.http_type}://#{config.fqdn}/#ticket/zoom/#{ticket.id}">ticket.id</a>}`

correct would be:

`#{ticket.id}`


@MrMontesa could you add your trigger perform content of your trigger to have an real example?

The output of  `zammad run rails r "p Trigger.find_by(name: 'your trigger name').perform"`

Thanks for feedback!

**Comment by MrMontesa** [comment]
Hey @martini 
Please find below the output of the requested command:

```
[root@tickets ~]# zammad run rails r "p Trigger.find_by(name: 'Ticket closing').perform"
{"notification.email"=>{"body"=>"<p>Your ticket (Ticket#\#{ticket.number}) is closed<br></p><p><br><br></p><p>To provide additional information, please on the following link:&nbsp;</p><p><br></p><p><span style=\"background-color: initial;\">\#{config.http_type}://\#{config.fqdn}/#ticket/zoom/\#{ticket.id}</span></p><p><br></p><p><br></p><div><br></div><div>Your \#{config.product_name} Team</div>", "recipient"=>"ticket_customer", "subject"=>"Ticket closed: \#{ticket.title}"}}
[root@tickets ~]#
```

Let me know if I can be of any help.
Thanks much
Sebastian

**Comment by martini** [comment]
@MrMontesa thanks for the fast feedback. It's good, this notification of trigger is different from that which I expected. 👍🏻

**Comment by ghost** [comment]
Same Problem here with the latest Chrome and Firefox Browser. The strange thing is that the problem exists only with trigger "ticket closed".  We have a second trigger for incoming notifications for the customer. And in these mail the url is working.

**Comment by rlue** [comment]
@MrMontesa / @qspweissen, what is the output of the following?

```
$ zammad run rails r "p Trigger.find_by(name: 'Ticket closing').condition"
```

**Comment by ghost** [comment]
i am running zammad on docker and these command didn't work:

/opt/zammad# rails r "p Trigger.find_by(name: 'Ticket closing')
I, [2018-07-10T11:33:06.893240 #126]  INFO -- : Setting.set('product_logo', "04a98b2eacc9495949a77a91c62a1d52.png")
Please specify a valid ruby command or the path of a script to run.
Run 'bin/rails runner -h' for help.

undefined method `condition' for nil:NilClass_

**Comment by rlue** [comment]
@qspweissen, looks like you're missing a trailing quotation mark?

**Comment by MrMontesa** [comment]
@rlue, I'm on vacation right now. Will follow up week after next once returned.
Thanks

**Comment by rlue** [comment]
@qspweissen, what version of Zammad are you running? And sorry, just realized that the `rails r` command I provided above only applies to @MrMontesa; could you modify it to reflect the name of the trigger that's giving you problems? (_i.e.,_ `Trigger.find_by(name: '<TRIGGER NAME>')`).

I'm not able to reproduce this issue; can anyone confirm that it happens on Zammad 2.5.0?

**Comment by ghost** [comment]
I am on version 2.5. 
`rails r "p Trigger.find_by(name: 'Ticket geschlossen').condition"`
Output:
```
INFO -- : Setting.set('product_logo', "04a98b2eacc9495949a77a91c62a1d52.png")
{"ticket.state_id"=>{"operator"=>"is", "value"=>"4"}}
```

`rails r "p Trigger.find_by(name: 'Ticket geschlossen').perform"`
Output:
```
INFO -- : Setting.set('product_logo', "04a98b2eacc9495949a77a91c62a1d52.png")
{"notification.email"=>{"body"=>"<p>Hallo&nbsp;\#{ticket.customer.firstname},</p><p><br></p><div></div><div>der Status deines Tickets \"\#{ticket.title}\" wurde von \#{ticket.updated_by.firstname}&nbsp;&nbsp;in \"<span style=\"background-color: initial;\">\#{</span><a href=\"http://ticket.state.name\" title=\"http://ticket.state.name\" target=\"_blank\">ticket.state.name</a><span style=\"background-color: initial;\">}\" geändert.</span></div><div></div><div><div><br></div><div>Das Ticket ist über folgenden Link erreichbar:&nbsp;<a href=\"https://ticket.xyz.de/\#{config.http_type}://\#{config.fqdn}/#ticket/zoom/\#{ticket.id}\">\#{config.http_type}://\#{config.fqdn}/#ticket/zoom/\#{ticket.id}</a></div><div><br></div><div><span style=\"background-color: initial;\">xyz IT Service Management</span></div></div>", "recipient"=>"ticket_customer", "subject"=>"\"\#{ticket.title}\" Geänderter Status \"\#{ticket.state.name}\""}}
```

The problem is the href link.

Here is my trigger:

```
Hallo #{ticket.customer.firstname},

der Status deines Tickets "#{ticket.title}" wurde von #{ticket.updated_by.firstname}  in "#{ticket.state.name}" geändert.

Das Ticket ist über folgenden Link erreichbar: #{config.http_type}://#{config.fqdn}/#ticket/zoom/#{ticket.id}
```

**Comment by martini** [comment]
JFI: I can reproduce it on FF and Crome:

1) open existing trigger
2) copy existing link into clipboard 
3) past link into same text, just 2 lines below
4) by using the browser element inspector you can see that the browser added the current hostname to the href attribute.

Workaround: Select the new inserted link and press ctrl+alt-x (see Zammad shortcuts) to remove hyperlinks (this also removed the damaged href attribute). Store it and it will send correct links again.

**Comment by rlue** [comment]
### EDIT

Whoops, looks like Martin beat me to the punch.

---

The only time I have been able to reproduce this issue is when pasting in rich text that contains hyperlinks, like so:

![issue_2019_rich_text_links](https://user-images.githubusercontent.com/12194123/42678788-8fafc65e-86b2-11e8-8c75-0b6dea73963a.png)

Based on your console output, it looks like @qspweissen's trigger suffers from this problem, whereas @MrMontesa's does not — which makes me wonder if they are in fact two different issues.

Could I ask you two to try removing any rich text hyperlinks from your trigger email bodies to see if that fixes the problem? There are a couple of ways to do this:

1. **Keyboard shortcut:** <kbd>Ctrl</kbd> + <kbd>Alt</kbd> + x should remove all links on Mac. Not sure if there is an equivalent shortcut on Windows / Linux.
2. **Manually:** delete the hyperlink portions of the trigger email body and re-type them.

This would really go a long way to helping us debug this problem. Thanks!

**Comment by rlue** [comment]
Just a quick update: we've got a fix in the works that should both fix existing triggers and ensure that future triggers do not suffer from this issue.

Links may still appear in (or be pasted into) the trigger editor, but duplicated domains will be stripped from the input when you click "submit".

---

@MrMontesa, [in the community forum, you said](https://community.zammad.org/t/malformed-url-links-in-triggered-mails/666/2):

> I analyzed app/views/mailer/ticket_update/en.html.erb and found... [code sample]
>
> This is exactly what I have in my trigger. **When replacing the href part with my own content it started working...**

Just to be sure I understand the issue, does this mean that

1. You found (and successfully implemented) a workaround at that time?
2. And that the trigger form looked like this before the workaround

   ![](https://community.zammad.org/uploads/default/original/1X/fed17c40a17ce7fc587788899cbc9e6b5996159d.png)

   and this after the workaround?

   ![](https://community.zammad.org/uploads/default/original/1X/2785ba71148ac0b644aa161cf31b40acbba0727c.png)
3. And thus, your [console output from May 18](https://github.com/zammad/zammad/issues/2019#issuecomment-390172374) was actually from **after** you fixed the bug, not before?

If all of the above is correct, that would explain why we don't see any unusual `href` values in your `Trigger#perform` output, when we do see it in @qspweissen's. (And if not, then there may be even more to this issue than I had originally assumed! :confused:)

**Comment by ghost** [comment]
The shortcut does not work for windows but i copy the link to VisualStudioCode and paste it again into zammad. After this the link works!
![2018-07-16 11_38_55-it helpdesk - ereignis basierte ausloser](https://user-images.githubusercontent.com/22341283/42752020-e4326934-88ec-11e8-8b5d-bfba84dda914.jpg)

**Comment by MrMontesa** [comment]
Hey there,
I'b back from vacation and happily saw some great feedback here. Unfortunately I lost track where I can help. One formation you asked for was the rails command.
We're on 2.4 right now and the commend executed with:
```
[root@tickets zammad]# zammad run rails r "p Trigger.find_by(name: 'Ticket closing').condition"
{"ticket.state_id"=>{"operator"=>"is", "value"=>"4"}}
```
I did **not** edit _app/views/mailer/ticket_update/en.html.erb_. This was only to confirm that I have the correct syntax of my link and some variable must have been filled wrong or similar.

If I remember correctly I worked around the issue by manually typing the link instead of copy/paste from  the initial trigger that was available.

Let me know if I can be of any help to get this sorted.
Thanks much for your efforts.
